### PR TITLE
NoUnusedImportsFixer - Fix undetected unused imports when type mismatch

### LINKED ISSUE #3020: Not removing unused import if has same name function
PHP CS Fixer not removing unused import that has the same name as that of the function used somewhere in class.

*Sample Code*

```php
<?php

namespace App\Controllers;

use App\Api\Controller;
use Zend\Diactoros\Exception;

class ValidationController extends Controller
{
    public function ipSubnet()
    {
        return exception('Something', 404);
    }
}
```

*Expected Code*

```php
<?php

namespace App\Controllers;

use App\Api\Controller;
// Remove unused import here

class ValidationController extends Controller
{
    public function ipSubnet()
    {
        return exception('Something', 404);
    }
}
```

The PHP version you are using:
$ php -v
```
PHP 7.1.9 (cli) (built: Aug 30 2017 18:58:30) ( NTS )
Copyright (c) 1997-2017 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2017 Zend Technologies
```

PHP CS Fixer version you are using:
$ php-cs-fixer -V
```
PHP CS Fixer 2.4.0 Silver Gingerbread by Fabien Potencier and Dariusz Ruminski (63661f3)
```

### PR DESCRIPTION
**Pull request by julienfalque**
Fixes #3020.

### CODE REVIEW

**Review by GrahamCampbell — APPROVED** [review]
Nice. 👍

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]
This incorrectly removes on of the imports for a union catch:

```
// ...
} catch (InvalidFormatException | ParseErrorException) {
```

**Comment by julienfalque** [comment]
@GrahamCampbell I don't reproduce that, could you provide the full snippet please?

**Comment by GrahamCampbell** [comment]
Oh, I think the issue only occurs when combined with https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5685.

**Comment by julienfalque** [comment]
@GrahamCampbell Actually I didn't notice your snippet uses a non-capturing catch 😐. I do reproduce the issue now, trying to fix it.

**Comment by julienfalque** [comment]
Fixed in #5685, which this PR now depends on. Thanks @GrahamCampbell!

**Comment by GrahamCampbell** [comment]
I think this also has an issue with PHP 8 attributes.

```php
<?php

use Symfony\Component\Routing\Annotation\Route;

class Controller
{
    #[Route('/foo', name: 'foo')]
    public function foo(Request $request): Response
    {
    }
}
```

**Comment by julienfalque** [comment]
@GrahamCampbell Fixed too, thanks :+1:

**Comment by GrahamCampbell** [comment]
Where there other changes there, or just a rebase?

**Comment by julienfalque** [comment]
Only a rebase (onto #5741 instead of #5685).

**Review by kubawerlos — APPROVED** [review]

**Comment by SpacePossum** [comment]
Fixed another issue for `goto` labels and in detecting `goto` labels. Please rebase and squash my commits out.
There is a thing I'm not 100% about:

```
                if ($analyzer->isConstantInvocation($index)) {
                    $type = NamespaceUseAnalysis::TYPE_CONSTANT;
                } elseif ($nextMeaningfulToken->equals('(') && !$prevMeaningfulToken->isGivenKind($tokensNotBeforeFunctionCall)) {
                    $type = NamespaceUseAnalysis::TYPE_FUNCTION;
                } else {
                    $type = NamespaceUseAnalysis::TYPE_CLASS;
                }
```

First case is 100% `constant` invocation. The second, `function` and 3rd (default) `class`, are not checked on the same detail level. If `T_STRING` would be used in the future for a new PHP feature we would default here to `class`, while previous to this PR the default was to assume it was used. Could we improve the testing without adding to much code?

**Comment by julienfalque** [comment]
> If T_STRING would be used in the future for a new PHP feature we would default here to class, while previous to this PR the default was to assume it was used. Could we improve the testing without adding to much code?

I think the current code is fine. If we know for sure that the tested token does not match a `function` or `const` import, then the token can only match a class/interface/trait import, which is what `NamespaceUseAnalysis::TYPE_CLASS` means. This would only fail if PHP introduces a new type of import but then supporting it would be part of the updates we do for each new PHP version.

**Comment by SpacePossum** [comment]
The `goto` case was also considered a "class" usage by the code because it is a T_STRING and is not a constant and not a function. I think when PHP would use the T_STRING token in the future for something different or new it would happen again.
For now I cannot think of such a case so we should be good though :)

**Comment by GrahamCampbell** [comment]
Merge conflicts. 😭

File: tests/Tokenizer/Analyzer/GotoLabelAnalyzerTest.php [hunk 1]
```diff
@@ -83,6 +83,17 @@ function A(){}
                 ',
                 [10],
             ],
+            'handle goto labels 5' => [
```
Comment (kubawerlos) [inline tests/Tokenizer/Analyzer/GotoLabelAnalyzerTest.php:86]: Can we drop the indices? The don't add any information.

**Review by kubawerlos — APPROVED** [review]

File: tests/Tokenizer/Analyzer/GotoLabelAnalyzerTest.php (hunk 1 above)
Comment (julienfalque) [reply tests/Tokenizer/Analyzer/GotoLabelAnalyzerTest.php]: Done.

**Review by GrahamCampbell — APPROVED** [review]

**Review by kubawerlos — APPROVED** [review]

**Comment by keradus** [comment]
Thank you @julienfalque.

**Comment by keradus** [comment]
thanks @SpacePossum

**Comment by andyrooster** [comment]
@julienfalque I think this changed has caused an issue with not detecting PHP 8 attributes usage when there is more than one in a block, for example with the below code the `IsGranted` and `JsonSchema` use statements get removed using CS Fixer v3.1.0. This doesn't happen in CS Fixer v3.0.2

```
use Sensio\Bundle\FrameworkExtraBundle\Configuration\IsGranted;
use Symfony\Component\Routing\Annotation\Route;
use Symfony\Component\Security\Core\Security;

...
#[
    Route('/basket/{uuid}/item', name: 'addBasketItem', requirements: ['uuid' => '%regex.uuid%'], methods: ['POST']),
    IsGranted('ROLE_USER'),
    JsonSchema('Public/Basket/addItem.json'),
]
```

**Comment by GrahamCampbell** [comment]
@andyrooster can you create a bug ticket for this following the template, so this can be properly tracked?

**Comment by andyrooster** [comment]
@GrahamCampbell https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5940

**Pull request merged**